File: android/app/src/com/android/bluetooth/le_scan/PeriodicScanNativeInterface.java
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.le_scan;

import static java.util.Objects.requireNonNull;

import android.bluetooth.BluetoothDevice;

import java.lang.annotation.Native;

public class PeriodicScanNativeInterface {
    private static final int PA_SOURCE_LOCAL = 1;
    private static final int PA_SOURCE_REMOTE = 2;

    @Native private final PeriodicScanNativeCallback mNativeCallback;

    PeriodicScanNativeInterface(PeriodicScanNativeCallback nativeCallback) {
        mNativeCallback = requireNonNull(nativeCallback);
    }

    void init() {
        initializeNative();
    }

    void cleanup() {
        cleanupNative();
    }

    void startSync(int sid, String address, int skip, int timeout, int regId) {
        startSyncNative(sid, address, skip, timeout, regId);
    }

    void stopSync(int syncHandle) {
        stopSyncNative(syncHandle);
    }

    void cancelSync(int sid, String address) {
        cancelSyncNative(sid, address);
    }

    void syncTransfer(BluetoothDevice bda, int serviceData, int syncHandle) {
        syncTransferNative(PA_SOURCE_REMOTE, bda.getAddress(), serviceData, syncHandle);
    }

    void transferSetInfo(BluetoothDevice bda, int serviceData, int advHandle) {
        transferSetInfoNative(PA_SOURCE_LOCAL, bda.getAddress(), serviceData, advHandle);
    }

    private native void initializeNative();

    private native void cleanupNative();

    private native void startSyncNative(int sid, String address, int skip, int timeout, int regId);

    private native void stopSyncNative(int syncHandle);

    private native void cancelSyncNative(int sid, String address);

    private native void syncTransferNative(
            int paSource, String address, int serviceData, int syncHandle);

    private native void transferSetInfoNative(
            int paSource, String address, int serviceData, int advHandle);
}

File: android/app/src/com/android/bluetooth/le_scan/PeriodicScanNativeInterface.kt
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.le_scan

import android.bluetooth.BluetoothDevice
import com.android.bluetooth.profile.NativeInterface

class PeriodicScanNativeInterface(nativeCallback: PeriodicScanNativeCallback) :
    NativeInterface<PeriodicScanNativeCallback>(nativeCallback) {

    fun init() {
        initializeNative()
    }

    override fun cleanup() {
        cleanupNative()
    }

    fun startSync(
        sid: Int,
        address: String,
        addressType: Int,
        skip: Int,
        timeout: Int,
        regId: Int,
    ) {
        startSyncNative(sid, address, addressType, skip, timeout, regId)
    }

    fun stopSync(syncHandle: Int) {
        stopSyncNative(syncHandle)
    }

    fun cancelSync(sid: Int, address: String) {
        cancelSyncNative(sid, address)
    }

    fun syncTransfer(bda: BluetoothDevice, serviceData: Int, syncHandle: Int) {
        syncTransferNative(PA_SOURCE_REMOTE, bda.address, serviceData, syncHandle)
    }

    fun transferSetInfo(bda: BluetoothDevice, serviceData: Int, advHandle: Int) {
        transferSetInfoNative(PA_SOURCE_LOCAL, bda.address, serviceData, advHandle)
    }

    private external fun initializeNative()

    private external fun cleanupNative()

    private external fun startSyncNative(
        sid: Int,
        address: String,
        addressType: Int,
        skip: Int,
        timeout: Int,
        regId: Int,
    )

    private external fun stopSyncNative(syncHandle: Int)

    private external fun cancelSyncNative(sid: Int, address: String)

    private external fun syncTransferNative(
        paSource: Int,
        address: String,
        serviceData: Int,
        syncHandle: Int,
    )

    private external fun transferSetInfoNative(
        paSource: Int,
        address: String,
        serviceData: Int,
        advHandle: Int,
    )

    companion object {
        private const val PA_SOURCE_LOCAL = 1
        private const val PA_SOURCE_REMOTE = 2
    }
}
